hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java

```diff
@@ -2499,14 +2535,36 @@ private static List<HoodieColumnRangeMetadata<Comparable>> getFileStatsRangeMeta
-                                                                                HoodieEngineContext engineContext,
-                                                                                HoodieTableMetaClient dataMetaClient,
-                                                                                HoodieMetadataConfig metadataConfig) {
+  public static HoodieData<HoodieRecord> convertMetadataToPartitionStatsRecords(HoodiePairData<String, List<HoodieColumnRangeMetadata<Comparable>>> columnRangeMetadataPartitionPair,
+                                                                                HoodieTableMetaClient dataMetaClient) {
+    return convertMetadataToPartitionStatsRecords(columnRangeMetadataPartitionPair.flatMapValues(List::iterator),
+        Option.empty(), isShouldScanColStatsForTightBound(dataMetaClient));
+  }
+
+  public static HoodieData<HoodieRecord> convertMetadataToPartitionStatsRecords(HoodiePairData<String, HoodieColumnRangeMetadata<Comparable>> columnRangeMetadataPartitionPair,
+                                                                                Option<String> indexPartitionOpt, boolean isTightBound) {
+    try {
+      return columnRangeMetadataPartitionPair
+          .groupByKey()
+          .map(pair -> {
+            final String partitionName = pair.getLeft();
+            return collectAndProcessColumnMetadata(pair.getRight(), partitionName, isTightBound, indexPartitionOpt);
+          })
+          .flatMap(recordStream -> recordStream.iterator());
+    } catch (Exception e) {
+      throw new HoodieException("Failed to generate column stats records for metadata table", e);
+    }
+  }
+
+  public static HoodiePairData<String, List<HoodieColumnRangeMetadata<Comparable>>> convertMetadataToPartitionStatsColumnRangeMetadata(HoodieCommitMetadata commitMetadata,
+                                                                                                                                       HoodieEngineContext engineContext,
+                                                                                                                                       HoodieTableMetaClient dataMetaClient,
+                                                                                                                                       HoodieTableMetadata tableMetadata,
+                                                                                                                                       HoodieMetadataConfig metadataConfig) {
     List<HoodieWriteStat> allWriteStats = commitMetadata.getPartitionToWriteStats().values().stream()
         .flatMap(Collection::stream).collect(Collectors.toList());
     if (allWriteStats.isEmpty()) {
-      return engineContext.emptyHoodieData();
+      return engineContext.emptyHoodieData().mapToPair(o -> Pair.of("", new ArrayList<>()));
```

Comment:
This is not an empty pair. Given that you're eventually flat-mapping only the values from the pair data, why not simply flatten and send HoodieData instead of pair from this method itself? I don't see the key of the pair being used anywhere later.

Reply:
After fusion the mapping is removed